### ISSUE TITLE
Code-sign static xcframeworks with resource bundles

### ISSUE THREAD

**Issue by i-pavlov** [tuist/tuist#5761]
### What problem or need do you have?

We have been using Tuist for a while already, and it successfully solves most of our problems. But there are some areas where we need to add some hacks to modify default tuist behavior. One of these areas is embedding static frameworks.

**The problem** 

We rely on precompiled xcframeworks to embed our 3rdParty dependencies, it helps to improve CI built time + some of the frameworks are just not available in any other ways.

These frameworks can be static or dynamic and some of them also have resource bundles inside. 

- For dynamic frameworks Tuist doing pretty well - it detects that the framework is dynamic and instead of just linking also embeds the framework into the top-level target (and enables code signing). in that case, the resource bundle can be properly accessed from the framework.
- In the case of static frameworks the current behavior is different. Since the framework is static - usually there is no need to embed it, just need to link it. And that's what Tuist is doing.

But if a static framework has a resource bundle inside, we are getting a problem. Tuist generates a project file that just links the framework and does not embed it. Without embedding - that resource bundle will not be copied into the app bundle, so the framework will not be able to access its resources.

**Workarounds I tried:**

- There is no way to configure xcframework to be embedded - the choice happens automatically under the hood based on framework type
- I tried to add one more copy files step, that copies the framework into the Frameworks directory. The framework will be embedded and it's even possible to access the framework bundle on debug builds. 
```swift
copyFiles: [
    .frameworks(name: "TestFramework", files: [
       .folderReference(path: "3rdParty/Deps/TestFramework.xcframework")
    ])
],
``` 
Unfortunately, this option doesn't allow me to code-sign the framework. As a result, all TestFlight and Firebase builds are not able to launch. That's kind of makes sense.
```
Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.Tjf3SY/extracted/TestApp.app/Frameworks/TestFramework.framework : 0xe800801c (No code signature found.)
```
So as a result none of the workarounds works for me and I'm out of ideas now.

Do you have any suggestions on how to resolve it?
I understand, that I can always migrate to a dynamic framework/push vendors to provide dynamic frameworks and get proper behavior out of the box, but having a lot of dynamic frameworks in the app seems like a bit not so good idea.

### Potential solution

**Suggested solution**
- It's already possible to create custom copy phases into the Frameworks folder. Also, in the generated project - it's possible to configure code-signing for these files. 
What if Tuist can give a bit more control to devs and make it possible to specify code-sign/not-code sign for these copied files?

- Also, as another option - it would be nice to be able to configure it on `TargetDependency.xcframework `level. But interface-wise it'll be a bit harder to set up, since it somehow should be defaulted to "always embed" for dynamic frameworks.

### macOS version

14.1.2

### Tuist version

3.37.0

### Xcode version

15.1.0

**Comment by pepicrft** [comment]
Thanks, @i-pavlov, for reporting this one. As you noted, Tuist tries to infer the XcodeProj implementation details of your project's graph. This is a case where we wrongly assume how things should be integrated. To provide the best advice on how to solve this, I'd need to understand how that `.xcframework` is structured internally. Would you mind sharing more details about the directory structure and its content? Or even better, share the XCFramework you use if it's public.

I'm thinking that if resources are conventionally organized inside the XCFramework, we could detect that and fix how Tuist integrates them. If not, we can expose an API for developers to indicate that the XCFramework has resources and, therefore, it should be integrated differently.

**Comment by i-pavlov** [comment]
Thanks @pepicrft , I cannot attach the same framework I'm trying to embed, but I created a minimal example of it:

Here it is: [StaticFramework.xcframework.zip](https://github.com/tuist/tuist/files/13833539/StaticFramework.xcframework.zip)

In this framework, **.bundle** is located at the root of the framework:
<img width="404" alt="Screenshot 2024-01-05 at 12 08 28 AM" src="https://github.com/tuist/tuist/assets/6316558/a6eaa645-8088-4fb1-bbea-ada9166e7de3">

Some other frameworks can ask to do the opposite: do not embed the framework, and manually embed .bundle into the app. For example - Adyen3DS2: https://github.com/Adyen/adyen-3ds2-ios#static-xcframework

In that case, automatic detection of such a thing can be tricky. I tried this one with my framework too:

```swift
copyFiles: [
  .frameworks(name: "Embed static frameworks", files: [
    .folderReference(path: "3rdParty/Deps/StaticFramework.xcframework/ios-arm64/StaticFramework.framework/com.framework.static-framework.bundle.bundle")
  ])
],
``` 
But it also doesn't work, project generation failed with an error:
```
Couldn't find a reference for the file at path /Users/ipavlov/Developer/client-ios/3rdParty/Deps/StaticFramework.xcframework.
Consider creating an issue using the following link: https://github.com/tuist/tuist/issues/new/choose
```

**Comment by kwridan** [comment]
Thanks @i-pavlov 

This sounds like a re-occurrence or a long standing issue https://github.com/tuist/tuist/issues/1779 however within XCFrameworks as opposed to plain old frameworks.

Embedding the XCFramework will include the binary along with the resource bundles of the XCFramework (which would be redundant). 

The solution for the plain frameworks was to specify the bundle as a resource, the main caveat there is in Tuist you can't reference a subpath a framework / XCFramework that is already in use, as such need to resort to symlinking the bundle outside of the XCFramework path and referencing that as a resource.

```swift
Target(
  // ...
  resources: [
     "path/to/symlinked.bundle",
 ]
)
```

**Comment by pepicrft** [comment]
> The solution for the plain frameworks was to specify the bundle as a resource, the main caveat there is in Tuist you can't reference a subpath of a framework / XCFramework that is already in use, as such need to resort to symlinking the bundle outside of the XCFramework path and referencing that as a resource.

@kwridan what do you think we extend `TargetDependency.xcframework` to support indicating resources that should be included as resources of the upstream dependent targets? Maybe something along the lines of:

```swift
TargetDependency.xcframework(..., resources: ["com.framework.static-framework.bundle.bundle"])
```

It'd require scripting to ensure the right resource is copied based on the platform, SDK, and architecture. Because as far as I know, this is not something supported by Xcode out of the box, isn't it?

**Comment by i-pavlov** [comment]
Yeah, from my perspective - I'd prefer to have bundles to be located inside the framework. Placing them as a single entity in the repository helps avoid possible inconsistencies, when someone updates the framework but forgets to update the bundle located nearby.

On pre-Tuist, with Xcode, it also wasn't possible to embed different bundles based on SDK so we ended up embedding the whole framework instead. But it requires code signing + since framework symbols are duplicated - it increases the binary size a bit.

**Comment by kwridan** [comment]
Complementing xcframework with an optional `resources` parameter could work. 

It could be a simple copy files build phase from `$(BUILT_PRODUCTS_DIR)/StaticFramework.framework/<resource>` which would already be the correct slice as Xcode would have processed the xcframework and extracted the correct framework if it's declared as link dependency. (That said, for mac frameworks resources are located in a different location, so maybe some script is needed).

The bit I don't have good clarity on how some of those static xcframeworks are built, would they be consistent / follow convention of having the resources within each framework of the xcframework?

From a cursory glance:

- In the example above, it's embedded within the framework 
- In [Firebase](https://github.com/firebase/firebase-ios-sdk/r), there's a different convention of an external `Resources` folder

**Comment by i-pavlov** [comment]
It's technically possible to pack frameworks in completely different ways. If we use dynamic frameworks as an example, where it happens a bit more often: 

- Mapbox just embeds resources directly into the framework without packing them into a bundle.
<img width="1068" alt="Screenshot 2024-01-05 at 4 49 26 PM" src="https://github.com/tuist/tuist/assets/6316558/56b38d33-b34e-47af-95aa-cdc2a71b88d8">

**Comment by pepicrft** [comment]
> It's technically possible to pack frameworks in completely different ways. If we use dynamic frameworks as an example, where it happens a bit more often:

We definitively need an API for developers to configure that. Would you like to take a stab at implementing this @i-pavlov. You should first try to do the setup in a standard Xcode project with a few XCFrameworks from the community. You can share the project here for @kwridan and I to provide input on what the setup should look like and make sure we don't miss any scenario.

Once we are happy with that, we can translate the setup to a set of APIs in Tuist and some logic to generate the necessary build phases and settings.

**Comment by i-pavlov** [comment]
Yep, it sounds like a plan. I'll find some xcframeworks from the community, so we'll be sure that we can cover most of the common use cases 👌

**Comment by pepicrft** [comment]
> Yep, it sounds like a plan. I'll find some xcframeworks from the community, so we'll be sure that we can cover most of the common use cases 👌

Sweet. Let us know when you have those.

**Comment by Ernest0-Production** [comment]
Is there any updates? :c 

I faced a similar problem and was forced to fork the vendor's dependency and extract resources to integrate through SPM. But I don't really want to do this with every update.

**Comment by i-pavlov** [comment]
In our case, we switched to the dynamic framework of the same vendor + updated the packaging behavior of our internal frameworks to ship resources separately and embed them as bundles.

While researching, I couldn't find many open-source / widely used examples of static frameworks with resource bundles inside. So, it's not clear to me if it's a best practice or not.

**Comment by vldalx** [comment]
Hey guys!
I'm trying to achieve a similar thing - distribute some dependence as a static frameworks with bundled resources.

My public framework is placed in a separated git repository. I would like to use `tuist` to generate the project file.
If I just set `product: .staticFramework` tuist will move all the resources from the framework target to a new bundle target. Building the project does not put the resource bundle into the framework because the **Copy Bundle Resource** build phase is empty.
Is it possible to make `tuist` adding the resource bundle to the **Copy Bundle Resource**?

Trying to find a solution, I found [this article](https://developer.apple.com/documentation/xcode/creating-a-static-framework). It stands that Xcode 15 supports static frameworks with bundled resources.
> In Xcode 15 or later, you can use a static framework to bundle resources with a static library

> You can add resources such as asset catalogs, storyboards, image files, and a privacy manifest to a static framework. The resource files appear in the Copy Bundle Resources build phase of the static framework target. 

Sounds prominent)

This makes me thinks, that the old approach with the bundle target can be deprecated now.
Unfortunately, the article doesn't say anything how to get resources from the static framework in code. Static frameworks do seem to [have issues](https://forums.developer.apple.com/forums/thread/751477) with it, but it can be worked around.

With this new approach (instead of bundle target) we can easily distribute static frameworks with bundled resources. No need to copy resources to the main target. In order to so with `tuist` we need to update it's logic: do not put resources in bundles and change generated `Bunde.module`.

What do you thing guys?

**Comment by fortmarek** [comment]
> Trying to find a solution, I found [this article](https://developer.apple.com/documentation/xcode/creating-a-static-framework). It stands that Xcode 15 supports static frameworks with bundled resources.

I wasn't aware of that change, thanks for linking that!

>  - (optionally) do not put resources in bundles
> - change generated `Bunde.module` to [get resources from the static framework](https://forums.developer.apple.com/forums/thread/751477?answerId=811319022#811319022)
> - static frameworks [needs to be embedded](https://developer.apple.com/documentation/xcode/creating-a-static-framework#Embed-your-framework-in-an-app) to the app target

Yes, that sounds like a reasonable approach! I'd try applying those steps manually in this [fixture](https://github.com/tuist/tuist/tree/main/fixtures/ios_app_with_static_frameworks_with_resources) and seeing if everything works. Then we can move on to the implementation.

If you end up trying that and things do work, it'd be great if you could add the sample here, so we can take a look at that as well 🤞

**Comment by vldalx** [comment]
> If you end up trying that and things do work, it'd be great if you could add the sample here, so we can take a look at that as well 🤞

here you go 
[StaticFrameworkApp.zip](https://github.com/user-attachments/files/17575752/StaticFrameworkApp.zip)

one non-obvious point:
When you use a static framework in your app, you still need to embed the framework so that you get all of its resources. However, what gets embedded is a cut down framework, with all the resources but only a stub Mach-O image.

**Comment by fortmarek** [comment]
Thanks!

This is really insightful. Then I'm onboard with moving away from generating a separate bundle since hosting resources by static frameworks is supported by Xcode.

Would you be up to implementing that? Happy to help in any way I can 🙂

**Comment by vldalx** [comment]
> Would you be up to implementing that?

At first glance, static frameworks now behave pretty much like dynamic ones. 
Thus, `Product.staticFramework` should be handled similarly to `Product.framework`, i.e. we just need to remove the bundle target creation step and code related to it.
Do you agree? Could you tell which classes need to be changed? This will help estimate the amount of work.

> if we do end up removing the behavior of us generating the resource bundle, we don't cause a breaking change in the current projects

It looks like re-creating a project using the updated `tuist` will be seamless if `Bundle.module` is used in codebase.
This won't work n Xcode 14 or older, of course. 

To be on the safe side, we can add a new `case` to `Product` enum and implement new logic behind it. Not sure if it's really needed.

**Comment by fortmarek** [comment]
> Thus, Product.staticFramework should be handled similarly to Product.framework, i.e. we just need to remove the bundle target creation step and code related to it.
Do you agree? Could you tell which classes need to be changed? This will help estimate the amount of work.

Yes, that makes sense to me 👍 

You'd need to change which targets support resources [here](https://github.com/tuist/XcodeGraph/blob/2f5f98a799b221c31c74f53b1ef8ff14c57b394a/Sources/XcodeGraph/Models/Target.swift#L223-L229). Once you do that, you should see if there's anything else missing by running tests and running generation locally but likely, only limited set of changes might be necessary. I assume we still want to keep the generation of resource targets for other targets like `.commandLineTool`, so not a lot of cleanup, such as removing `ResourcesProjectMapper` should not be necessary.

>  This won't work n Xcode 14 or older, of course.

We don't support Xcode 14, so we should be safe.

**Comment by pbk20191** [comment]
As far as I know, Resource Bundle can not be unified like xcframework does, 
making Resource bundle can be tricky. If the resource bundle contains Metal.library.

each platform produce different Metal.library from same code, and Metal library act like resource.
So I'm not sure how tuist can handle the case like below.

1. Swift Package that contains Metal Library and code with static framework target
2. fetch the package and make a `cache`
3. we have one universal xcframework and N number of metal.lib bundles?


example) https://github.com/EsotericSoftware/spine-runtimes
